### ISSUE TITLE
Release 0.10.0

### ISSUE THREAD

**Issue by bacherfl** [keptn/keptn#5462]
# Keptn 0.10.0

* **Link to milestone(s): [Milestone 0.10.0](https://github.com/keptn/keptn/milestone/0.10.0)** 
* **Release Date: <= October, 2021** 
* **Release Managers:** @bacherfl @mowies 

### Step 1 - Release of keptn/spec [OPTIONAL]

**In case no new keptn/spec is released:**
- [x] This version of Keptn does not require a new version of keptn/spec. 

**else:**notification

* **Artifact:** GitHub Release `keptn/spec:$KEPTN_SPEC_VERSION` (also referred as ``spec.keptn.sh/$KEPTN_SPEC_VERSION``)
* **Assignee**: Firstname Lastname (insert GitHub handle here) (**TODO: Assign this step to someone**)
* **Repo**: https://github.com/keptn/spec
* **Note**: Not every Keptn release requires a new Keptn spec.

**Tasks:**
- [x] All release-related PRs/discussions are resolved and are merged into master (please check back with the Issue/PR owner)
- [x] keptn/go-utils is in sync with spec/cloudevents (e.g., using `keptn generate cloud-events`)
- [x] **Cleanup (Requires ADMIN role):** Inactive branches are deleted
- [x] *Draft* a new release $KEPTN_SPEC_VERSION https://github.com/keptn/spec/releases with release notes (collect them from PR/discussion titles) on master branch - again, ensure a semantic version is used!
- [x] Ask for feedback about release notes via Slack (link to the draft release can only be seen by users with write access to the repo)
- [x] If you're happy with everything and the release notes have been approved/accepted, pull the trigger and release the spec
- [x] Notify Developers on Slack that a new release of keptn/spec has been created and link it (**AUTOMATION WANTED**)

---
### Step 2 - Release of keptn/examples

* **Artifact:** GitHub Release `keptn/examples:$KEPTN_VERSION`
* **Assignee**: Firstname Lastname (insert GitHub handle here) (**TODO: Assign this step to someone**)
* **Repo**: https://github.com/keptn/examples
* **Note**: Use the same version as for keptn/keptn (every Keptn release requires a new release of Keptn examples)

- [x] All release-related feature/bugfix/hotfix branches are merged into master (please check back with the PR owner)
- [x] **Cleanup (Requires ADMIN role):** Inactive branches are deleted
- [x] Created `release-$KEPTN_VERSION` branch: `git checkout -b release-$KEPTN_VERSION` - ensure semantic version is used
- [x] _on release-$KEPTN_VERSION branch:_ releasenotes_develop.md renamed to releasenotes_**V$KEPTN_VERSION**.md (**AUTOMATION WANTED**)
- [x] _on release-$KEPTN_VERSION branch:_ collect release-notes either based on GitHub Issues for this Release or via commit messages (**AUTOMATION WANTED**)
- [x] _on release-$KEPTN_VERSION branch:_ Update README.md with links to *Maintained Examples and Keptn Use cases* (not necessary for pre-releases)
- [x] _on release-$KEPTN_VERSION branch:_ Don't forget: `git commit`, `git push [-u origin release-$KEPTN_VERSION]`
- [x] Create PR from release-$KEPTN_VERSION to master and request reviews (review of release notes is mandatory)
- [x] If the PR is mergeable and approved, get it merged into master (**don't squash merge**) 
- [x] *Draft* a new release https://github.com/keptn/examples/releases with the release notes on the `release-$KEPTN_VERSION` branch (not on **master**) (**AUTOMATION WANTED**)
- [x] If you're happy with everything, pull the trigger and release the examples
- [x] Notify Developers on Slack that a new release has been created (**AUTOMATION WANTED**)
- [x] Created new and empty release notes (releasenotes_develop.md) file on master branch

--- 
### Step 3 - Release of Keptn kubernetes-utils

* **Artifact:** GitHub Release of `keptn/kubernetes-utils:v$KEPTN_VERSION` 
* **Assignee**: Firstname Lastname (insert GitHub handle here) (**TODO: Assign this step to someone**)
* **Repo**: https://github.com/keptn/kubernetes-utils
* **Note:** The `v` in the version, which can be used in GoLang using `go get github.com/keptn/kubernetes-utils@v$KEPTN_VERSION`

- [x] All release-related feature/bugfix/hotfix branches are merged into master (please check back with the PR owner)
- [x] **Cleanup (Requires ADMIN role):** Inactive branches are deleted
- [x] If this is a fix/patch release and a maintenance branch (`0.N.x`) isn't there yet for the current feature/minor version, create the according branch. (for release of next bug version after `v0.8.6` use the branch `0.8.x`)
  If this is an arbitrary release with other features, then the release should not happen on a maintenance branch but rather from master directly to automatically create a new feature/minor version.
- [x] If you're happy with everything, pull the trigger and release keptn/kubernetes-utils, use the 'Release' GitHub action and trigger a manual run with the branch that you want to release from. (❗️ **CAUTION: this will create the release commit and tag!** ❗️)
- [x] The GitHub release was created in draft mode. Double check the content of the new release and then publish it if everything is correct.
- [x] If the release was done from a maintenance branch, create a PR from the released branch to master and request reviews (review of release notes is mandatory)
- [x] If the PR is mergeable and approved, get it squash & merged into master (**don't delete the branch if it was a maintenance release, restore the branch if GitHub automatically deleted it!**)
- [x] Notify Developers on Slack that a new release has been created (**AUTOMATION WANTED**)

--- 
### Step 4 - Release of Keptn go-utils

* **Artifact:** GitHub Release of `keptn/go-utils:v$KEPTN_VERSION` 
* **Assignee**: Firstname Lastname (insert GitHub handle here) (**TODO: Assign this step to someone**)
* **Repo**: https://github.com/keptn/go-utils
* **Note:** The `v` in the version, which can be used in GoLang using `go get github.com/keptn/go-utils@v$KEPTN_VERSION`

- [x] All release-related feature/bugfix/hotfix branches are merged into master (please check back with the PR owner)
- [x] **Cleanup (Requires ADMIN role):** Inactive branches are deleted
- [x] If this is a fix/patch release and a maintenance branch (`0.N.x`) isn't there yet for the current feature/minor version, create the according branch. (for release of next bug version after `v0.8.6` use the branch `0.8.x`)
  If this is an arbitrary release with other features, then the release should not happen on a maintenance branch but rather from master directly to automatically create a new feature/minor version.
- [x] If you're happy with everything, pull the trigger and release keptn/go-utils, use the 'Release' GitHub action and trigger a manual run with the branch that you want to release from. (❗️ **CAUTION: this will create the release commit and tag!** ❗️)
- [x] The GitHub release was created in draft mode. Double check the content of the new release and then publish it if everything is correct.
- [x] If the release was done from a maintenance branch, create a PR from the released branch to master and request reviews (review of release notes is mandatory)
- [x] If the PR is mergeable and approved, get it squash & merged into master (**don't delete the branch if it was a maintenance release, restore the branch if GitHub automatically deleted it!**)
- [x] Notify Developers on Slack that a new release has been created (**AUTOMATION WANTED**)

---

## Prepare release of associated services

:warning: **WARNING/STOP/DANGERZONE**: Do not start with this section unless **all the above steps** have been completed successfully.

---
### Step 5 - Create Release Branch for keptn-contrib services

**Please update version tags of below services accordingly; use semantic versioning**

**foreach:**
- [ ] prometheus-service > **prometheus-service:x.x.x**
- [ ] argo-service > **argo-service:x.x.x**
- [ ] servicenow-service > **servicenow-service:x.x.x**
- [ ] unleash-service > **unleash-service:x.x.x**

follow https://github.com/keptn/keptn/wiki/Keptn-Contrib-Release-Checklist Step 1

## Prepare Release of Keptn

---
### Step 6 - Release of Keptn
- [x] Update JMeter readme to reference $KEPTN_VERSION
- [x] `keptn upgrade` - verify if any changes/migrations/... in the CLI command are needed for this release
- [x] `keptn upgrade` - Create a PR at https://github.com/keptn/get.keptn.sh and modify [version.json](https://github.com/keptn/get.keptn.sh/blob/master/src/version.json) accordingly
- [x] Upgrade docs available in https://keptn.sh/docs/
- [x] Were there changes on K8s roles made? If yes, ensure (automatically) that the least privileges are always enforced for Keptn K8s roles: Who has which rights?
- [x] Ensure that the latest auto-generated PR to update the git submodule spec is merged in keptn/keptn. This PR is there to update the specification submodule when a new version of the spec is released.
- [x] All relevant feature/bugfix/hotfix branches are merged into master
- [x] **Cleanup (Requires ADMIN role):** Inactive branches are deleted
- [x] Latest go-utils version is used (run: `go get github.com/keptn/go-utils@v$KEPTN_VERSION && go get ./... && go mod tidy`) in all sub-folders that use it (see [this code snippet](https://gist.github.com/thisthat/c6b9018290ef3faeaa5c32ed39dc2fa8)) and the latest kubernetes-utils version is used (run: `go get github.com/keptn/kubernetes-utils@v$KEPTN_VERSION && go get ./... && go mod tidy`) in all sub-folders that use it (see [this code snippet](https://gist.github.com/thisthat/0431369bb1e1e747bc36350c1263e931)). 
- [x] Keptn release notes are up-to-date in `releasenotes/develop.md` and spec version is referenced in release notes 
- [x] Use [Create Release Branch Workflow](https://github.com/keptn/keptn/actions/workflows/create_release_branch.yml) Workflow
- [x] Verify `release-$KEPTN_VERSION` and a PR have been created (automated using "Create Release Branch Workflow")
- [ ] Ask stakeholders to review the release notes (*Note:* Do not use a GitHub DRAFT Release, it does not allow editing from multiple parties at the same time and we will lose changes**)
- [ ] Check [GH Workflow CI for release branch](https://github.com/keptn/keptn/actions?query=workflow%3ACI+branch%3Arelease-$KEPTN_VERSION)
- [ ] Check [GH Workflow Integration Test for release branch](https://github.com/keptn/keptn/actions?query=workflow%3A%22Integration+Tests%22+branch%3Arelease-$KEPTN_VERSION)
- [ ] Check for a draft release with the correct version https://github.com/keptn/keptn/releases (**DO NOT RELEASE IT YET**)
- [ ] Draft release contains CLI binaries for Linux, Windows and MacOS
- [ ] Draft release contains Keptn installer helm chart
- [ ] Draft release contains Helm Charts for jmeter-service and helm-service

Check on GCP Cloud Storage:

- [ ] keptn-installer $KEPTN_VERSION + index files updated

Notify certain users and stakeholders about the release, such that they can test it and ensure compatibility with their services/docs. Make sure to provide instructions on how to download the CLI and install Keptn. **Make sure to let them know that this might not the final release, rather that it is a release candidate.**

- [ ] Notify developers, keptn/tutorials CODEOWNERS, keptn-contrib and keptn-sandbox repo owners via Keptn Slack (e.g., keptn-integrations channel) about the upcoming release and provide a preview to the release notes

---
### Step 7 - Final Preparation phase

- [ ] keptn.sh/docs/integrations/ is up-to-date
- [ ] keptn CLI command docs generated and updated on keptn.sh/docs/x.x.x/reference/cli/commands/ (**TODO: Update link to docs**)
      A PR with the updated CLI command docs was created after the tag for keptn/keptn was pushed. You will find the PR in the keptn/keptn.github.io repo.
- [ ] keptn.sh/docs/ is up-to-date
- [ ] tutorials.keptn.sh: update of tutorials in this branch/PR: xyz (**TODO: Add link to PR/branch**)
  * Copy all 00.md files in site/tutorials and rename
  * Change tags and id in the .md files accordingly
  * Change versions 00 and 0.0 in the .md files
  * Copy all 00 folders in site/tutorials and rename  
  * Copy site/app/views/keptn00x folder
  * Edit site/app/views/keptn00x/view.json
  * Copy site/tutorials/assets/…/0.0 where needed
  * Run builder.sh

---
### Step 8 - Test phase

Upgrader:
- [ ] Upgrade command tested for full installation (GKE or K3s)
- [ ] Upgrade command tested for quality-gates/control-plane installation (GKE or K3s)
- [ ] Upgrade command tested for full installation (OpenShift 3.11 or MiniShift)
- [ ] Upgrade command tested for quality-gates/control-plane installation (OpenShift 3.11 or MiniShift)

Tests:
- Download of **keptn CLI $KEPTN_VERSION** > [ToDo: LINK to CLI]()
- Install keptn cli and verify that it prints the correct version: `keptn version`
- For keptn-contrib services: follow https://github.com/keptn/keptn/wiki/Keptn-Contrib-Release-Checklist Step 2
- Pre-view of **keptn.sh/docs** > [ToDo: LINK]()

**Note**: The following is partially covered by integration tests. For pre-releases, it can be skipped.
However, for GA releases the tests in this checklist need to be executed manually. Focus on usability and how it feels for the end-user!

- [ ] GCP - GKE - **Keptn core team**
  - [ ] installation: `keptn install`
    - [ ] Quality Gates UC completed according to tutorials.keptn.sh with Dynatrace
    - [ ] Quality Gates UC completed according to tutorials.keptn.sh with Prometheus
  - [ ] Installation: `keptn install --use-case=continuous-delivery`
    - [ ] Tutorials completed with [Dynatrace](https://tutorials.keptn.sh/tutorials/keptn-full-tour-dynatrace-08/index.html#0)
        - [ ] Onboarding a Service
        - [ ] Deployments with Quality Gates
        - [ ] Up-scale with Dynatrace
        - [ ] Self-healing with Feature Flags
        - [ ] Are there changes required on the Tutorial? > Please file a PR
    - [ ] Tutorials completed with [Prometheus](https://tutorials.keptn.sh/tutorials/keptn-full-tour-prometheus-08/index.html#0)
        - [ ] Onboarding a Service
        - [ ] Deployments with Quality Gates
        - [ ] Up-scale with Prometheus
        - [ ] Are there changes required on the Tutorial? > Please file a PR
  - [ ] Uninstall Keptn from cluster: `keptn uninstall`
  - [ ] Update https://keptn.sh/docs/x.x.x/operate/install/ based on used K8s version
- [ ] Azure - AKS
  - [ ] Installation: `keptn install --use-case=continuous-delivery`
  - [ ] Ensure `keptn version` prints the correct version
  - [ ] `keptn create project xyz`
  - [ ] Ensure Keptn Bridge can be accessed and works
  - [ ] Uninstall Keptn from cluster: `keptn uninstall`
  - [ ] Update https://keptn.sh/docs/x.x.x/operate/install/ based on used K8s version
- [ ] AWS - EKS
  - [ ] All tutorials completed with Dynatrace
      - [ ] Onboarding a Service
      - [ ] Deployments with Quality Gates
      - [ ] Up-scale with Dynatrace
      - [ ] Self-healing with Feature Flags
  - [ ] Update https://keptn.sh/docs/x.x.x/operate/install/ based on used K8s version
- [ ] OpenShift 3.11 - MiniShift
  - [ ] Update https://keptn.sh/docs/x.x.x/operate/install/ based on used OpenShift version
- [ ] OpenShift 4.x
  - [ ] All tutorials completed with Dynatrace
      - [ ] Onboarding a Service
      - [ ] Deployments with Quality Gates
      - [ ] Self-healing with Feature Flags
  - [ ] Update https://keptn.sh/docs/x.x.x/operate/install/ based on used OpenShift version
  - [ ] Update support matrix on keptn.sh/docs and release notes (key announcement + special thanks to Marc)
- [ ] Kubernetes (e.g., K3s, Rancher)
  - [ ] Installation: `keptn install --use-case=continuous-delivery`
  - [ ] Ensure `keptn version` prints the correct version
  - [ ] `keptn create project xyz`
  - [ ] Ensure Keptn Bridge can be accessed and works
  - [ ] Uninstall Keptn from cluster: `keptn uninstall`
  - [ ] Update https://keptn.sh/docs/x.x.x/operate/install/ based on used K8s version

---
### Step 9 - Official release of keptn-contrib services

**foreach:**
- [ ] dynatrace-service > **dynatrace-service:x.x.x**
- [ ] dynatrace-sli-service > **dynatrace-sli-service:x.x.x**
- [ ] prometheus-service > **prometheus-service:x.x.x**
- [ ] argo-service > **argo-service:x.x.x**
- [ ] servicenow-service > **servicenow-service::x.x.x**
- [ ] unleash-service > **unleash-service:x.x.x**

follow https://github.com/keptn/keptn/wiki/Keptn-Contrib-Release-Checklist Step 3

---
### Step 10 - Official release of Keptn

**Note**: Some steps are only for GA releases, not for alpha, beta, etc.
- [ ] Merge the PR at https://github.com/keptn/get.keptn.sh [version.json](https://github.com/keptn/get.keptn.sh/blob/master/src/version.json)

Check on GCP Cloud Storage:

- [ ] keptn-installer x.x.x

Release Keptn
- [ ] Publish the draft release for Keptn $KEPTN_VERSION on GitHub, select `release-$KEPTN_VERSION` as a target
- [ ] Verify that installing Keptn CLI works via `curl -sL https://get.keptn.sh | KEPTN_VERSION=$KEPTN_VERSION bash` and execute `keptn version` to check the version number
- [ ] Merge PR from **release-x.x.x** to **master** (**merge only if needed**)
- [ ] Created new and empty release notes (releasenotes_develop.md) file on master branch
- [ ] Update **Quick Start** on keptn.sh with newest release (**only for GA, not for pre-releases**)
- [ ] Add release notes to: https://keptn.sh/docs/news/release_announcements/ (**only for GA, not for pre-releases; AUTOMATION WANTED**)
- [ ] Update: https://github.com/keptn/keptn.github.io/blob/master/layouts/partials/announcement.html
---
### Final steps

- [ ] Send out Slack notification
- [ ] Send out Twitter notification
- [ ] Send out Release-Email
- [ ] Make sure any (relevant) changes to the release checklist are also saved in the [wiki](https://github.com/keptn/keptn/wiki/Release-Checklist)
